### ISSUE TITLE
Results Mismatch between XGBoost Model and ONNX Runtime when eval_set parameter is passed in model.fit method

### ISSUE THREAD

**Issue by Gunjitbedi** [microsoft/onnxruntime#5450]
**Describe the bug**
Prediction probabilities of an XGBoost model are different from ONNX Runtime when the same test input is passed through them. On debugging of the issue I found out that the miss-match happens when eval_set parameter is passed through model.fit method as shown below. The results match when this parameter is removed.

clr = XGBClassifier()
clr.fit(X_train, y_train, eval_set=[(X_test, y_test)], early_stopping_rounds=40)

**Urgency**
This bug does not allow us to train an XGboost model.

**System information**
- Windows 10
- ONNX Runtime installed from (source or binary): PIP
- ONNX Runtime version: 1.5.1
- Python version: 3.6
- Visual Studio version (if applicable): No
- GCC/Compiler version (if compiling from source): No
- CUDA/cuDNN version: CPU Only
- GPU model and memory: CPU Only

**To Reproduce**

```
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
import onnx
import onnxmltools
from xgboost import XGBClassifier
from skl2onnx.common.data_types import FloatTensorType

# Train a XGBoost model.
iris = load_iris()
X, y = iris.data, iris.target
X_train, X_test, y_train, y_test = train_test_split(X, y)
clr = XGBClassifier()

clr.fit(X_train, y_train, eval_set=[(X_test, y_test)], early_stopping_rounds=40)  # Passing eval_set parameter makes the results mismatch between XGBoost model and ONNX Runtime

# clr.fit(X_train, y_train) # When this is called results match between XGBoost model and ONNX Runtime

# Convert into ONNX format
initial_type = [('float_input', FloatTensorType([None, 4]))]
onx = onnxmltools.convert.convert_xgboost(clr, initial_types=initial_type)
with open("rf_iris.onnx", "wb") as f:
    f.write(onx.SerializeToString())

# ONNX Runtime
import onnxruntime as rt
import numpy as np
sess = rt.InferenceSession("rf_iris.onnx")
input_name = sess.get_inputs()[0].name
label_name = sess.get_outputs()[1].name

# Sample Test Input
predict_list = [1.,  20., 466.,   0.]
predict_array = np.array(predict_list).reshape((1,-1))

# Prediction from ONNX Runtime
pred_onx = sess.run([label_name], {input_name: predict_array.astype(np.float32)})[0]
print("Prediction of ONNX Runtime", pred_onx)

# Prediction from XBOOST Model
pred_xgboost = sessresults=clr.predict_proba(predict_array)
print("Prediction of XGBOOST MODEL", pred_xgboost)

```

**Expected behavior**
The results of both a trained XGboost model and its ONNX version should match

**Additional context**
For this example of sake, I have passed the Test data set as an evaluation set while fitting the model. Though I understand a validation set should have been passed. Although, the issue mentioned above remains the same

**Comment by xadupre** [comment]
I was able to replicate the issue. The bug happens when the model includes more than 1 tree. I'll be working on it.

**Comment by xadupre** [comment]
This issue comes from the converter, it does not check Booster.best_ntree_limit. That parameter is used to compute the prediction only with the first trees. This parameter is not checked at conversion time. This is fixed by PR https://github.com/onnx/onnxmltools/pull/429.

**Comment by wbu-wish** [comment]
Any luck on the status of this fix? I tried to install the latest onnxmltools with source but the problem still remains.

**Comment by rashmibadole** [comment]
@ 
Member
xadupre  as said by wbu-wish, I am too facing the same issue with onnxmltools 1.8.0 . Lets take the above iris data program. The results are not consistent.  I got the same probabilities for  xgboost and onnx a few times but then it starts giving different probabilities and even weird probabilties that do not add up to 1.0 ( for above eg.  Prediction of ONNX Runtime [[0.08729726 0.5        0.5       ]]). I see this issue only for classification. For regression using xgboost, the onnx and xgboost predictions match.  I am using onnxruntime version 1.8.1  .

**Comment by IsmaelGSerrano** [comment]
I am facing the same issue with XGBoostClassifier
Linux
Python 3.7

numpy: 1.19.5
scikit-learn: 0.24.2
onnx:  1.10.1
onnxruntime:  1.8.1
skl2onnx:  1.9.2
onnxmltools:  1.9.1
xgboost:  1.4.0

An example:
predict_onnx [1 0 1 0 1 1 0 0 0 0]
predict_proba_onnx [{0: 0.4202345609664917, 1: 0.5797654390335083}]
pipe results:
predict [1. 1. 1. 1. 1. 1. 0. 1. 0. 0.]
predict_proba [[0.29471636 0.70528364]]

pipe: [('scaler', StandardScaler()), ('model', XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.8075744610614244,
              eval_metric='logloss', gamma=0.31202049261817516, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.0733829737841539, max_delta_step=0, max_depth=18,
              min_child_weight=12.0, missing=nan, monotone_constraints='()',
              n_estimators=500, n_jobs=12, num_parallel_tree=1, random_state=0,
              reg_alpha=10.017604431945463, reg_lambda=27.53914730083718,
              scale_pos_weight=1, subsample=0.900023902289431,
              tree_method='hist', validate_parameters=1, verbosity=None))],

**Comment by xiaohk** [comment]
Have the same issue. Any update?

**Comment by fhaque** [comment]
Bump. Any update?
